url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/Compare/buildBtreeOfPoints.java

```diff
@@ -0,0 +1,186 @@
+					point[1] = Float.parseFloat(num[3]);
+					if(ISCAR){
+						point[0] = (int)(point[0] * SCALE);
+						point[1] = (int)(point[1] * SCALE);
+					}
+					points.add(point);
+					//if(DEBUG) System.err.println();
+				}
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+		} catch (FileNotFoundException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		final long start = System.currentTimeMillis();
+		delaunay = new Delaunay(points.toArray(new float[0][0]));		
+		final int[] lock = new int[1];
+		lock[0] = 0;
+		final int totalNum = points.size();
+		for(int id = 0; id < ThreadNum; id ++){
+			threadStatus[id] = false;
+			final int tid = id;
+			new Thread(new Runnable() {
+				int threadId;
+				@Override
+				public void run() {
+					// TODO Auto-generated method stub
+					threadId = tid;
+					while(true){
+						int curId;
+						synchronized (lock) {
+							curId = lock[0];
+							lock[0] ++;
+						}
+						if(curId >= totalNum)break;
+						System.out.println("Thread:\t" + threadId + "\tid:\t" + curId);
+						int[] pid = delaunay.getLinked(curId + 1);
+						ArrayList<Long> tmp = new ArrayList<Long>();
+						Point pPoint = new Point((int)points.get(curId)[0], (int)points.get(curId)[1], (int)points.get(curId)[2]);
+						for(int j = 0; j < pid.length && pid[j] > 0; j++){
+							tmp.add((long)idmap.get(pid[j] - 1));
+						}
+						pPoint.buildByPaillier();
+						DataOfPoint data = new DataOfPoint(curId, pPoint, tmp.toArray(new Long[0]));
+						data.signWithRSA(rsa);
+						try {
+							btOfPoint.put((long)idmap.get(curId), data.writeToBytes());
+						} catch (IOException e) {
+							// TODO Auto-generated catch block
+							e.printStackTrace();
+						}
+						if(curId % 500 == 0){
+							System.out.println("Commit Thread:\t" + threadId + "\tid:\t" + curId);
+							long end = System.currentTimeMillis();
+							long expectedTime = (end - start) / (curId + 1) * (totalNum - curId) / 1000;
+							System.out.println("Expected time remaining: " + expectedTime + " s = " + (expectedTime / 60) + " m = " + (expectedTime / 3600) + " h");
+							try {
+								recmanOfPoint.commit();
+							} catch (IOException e) {
+								// TODO Auto-generated catch block
+								e.printStackTrace();
+							}
+						}
+					}
+					threadStatus[threadId] = true;
+				}
+			}).start();
+		}
+		//pw.flush();
+		//pw.close();
+		while(true){
+			boolean found = false;
+			try {
+				Thread.sleep(100);
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+			for(int i = 0; i < ThreadNum; i++){
+				if(threadStatus[i] == false){
+					found = true;
+				}
+			}
+			if(!found)break;
+		}
+		recmanOfPoint.commit();
+	}
+	
+	public buildBtreeOfPoints(boolean needLoad, String sourceFileName, String destFileName) throws IOException{
+		recmanOfPoint = RecordManagerFactory.createRecordManager(destFileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource that might leak to a class field. This can cause a resource leak that slows down or crashes your system. For example, when buildBtreeOfPoints.loadData throws an exception, and an if condition at line 142 evaluates to true, the *buildBtreeOfPoints* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *recmanOfPoint* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).